ee/clickhouse/queries/paths/path_event_query.py

```diff
@@ -28,10 +35,18 @@ def get_query(self) -> Tuple[str, Dict[str, Any]]:
         funnel_paths_join = ""
         funnel_paths_filter = ""
 
-        if self._filter.funnel_paths:
-            funnel_paths_timestamp = f"{self.FUNNEL_PERSONS_ALIAS}.timestamp as min_timestamp"
+        if self._filter.funnel_paths == FUNNEL_PATH_AFTER_STEP or self._filter.funnel_paths == FUNNEL_PATH_BEFORE_STEP:
+            # used when looking for paths up to a dropoff point to account for events happening between the latest even and when the person is deemed dropped off
+            funnel_window = f"+ INTERVAL {self._filter.funnel_window_interval or self._filter.funnel_window_days} {self._filter.funnel_window_interval_unit_ch()}"
+            operator = ">=" if self._filter.funnel_paths == FUNNEL_PATH_AFTER_STEP else "<="
+
+            funnel_paths_timestamp = f"{self.FUNNEL_PERSONS_ALIAS}.timestamp AS target_timestamp"
+            funnel_paths_join = f"JOIN {self.FUNNEL_PERSONS_ALIAS} ON {self.FUNNEL_PERSONS_ALIAS}.person_id = {self.DISTINCT_ID_TABLE_ALIAS}.person_id"
+            funnel_paths_filter = f"AND {self.EVENT_TABLE_ALIAS}.timestamp {operator} target_timestamp {funnel_window if self._filter.funnel_paths == FUNNEL_PATH_BEFORE_STEP and self._filter.funnel_step and self._filter.funnel_step < 0 else ''}"
```

Comment:
Q: Why do we have the `funnel_window` on `FUNNEL_PATH_BEFORE_STEP`?

For `FUNNEL_PATH_BEFORE_STEP` don't we want to look at `timestamp <= target_timestamp` ? ( and maybe `timestamp >= target_timestamp - <INTERVAL>` )

And for `FUNNEL_PATH_AFTER_STEP`, `timestamp >= target_timestamp AND timestamp <= target_timestamp + <INTERVAL>`

Reply:
for after step, the interval shouldn't matter because we don't care about the funnel window when we're looking at paths. What we have to do here is actually make sure to convert `funnel_path_window` into `session_time_threshold`

For before step, we also don't need the intervals (as long as we use the right session_time_threshold like above) except for the case where you're asking for paths before a dropoff. A person is determined to have dropped off if the interval time has elapsed after their latest funnel step performed. So for the purposes of seeing how people got to a dropoff point we actually need to consider everything before target_timestamp + {interval} (because the target_timestamp we have here is only the timestamp of a person's latest event in the funnel

Reply:
Agreed about the `session_time_threshold`, makes sense.

>  So for the purposes of seeing how people got to a dropoff point we actually need to consider everything before target_timestamp + {interval} (because the target_timestamp we have here is only the timestamp of a person's latest event in the funnel

Oh, this is clever, makes sense, I didn't realise this. They could've done more events before dropping off, after the funnel step.